File: app/src/main/java/com/elementary/tasks/core/controller/ShoppingEvent.java
package com.elementary.tasks.core.controller;

import android.content.Context;
import android.text.TextUtils;

import com.elementary.tasks.core.utils.TimeCount;
import com.elementary.tasks.core.data.models.Reminder;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class ShoppingEvent extends RepeatableEventManager {

    ShoppingEvent(Reminder reminder, Context context) {
        super(reminder, context);
    }

    @Override
    public boolean start() {
        getReminder().setActive(true);
        super.save();
        if (!TextUtils.isEmpty(getReminder().getEventTime()) && TimeCount.isCurrent(getReminder().getEventTime())) {
            super.enableReminder();
        }
        return true;
    }

    @Override
    public boolean skip() {
        return false;
    }

    @Override
    public boolean next() {
        return stop();
    }

    @Override
    public boolean onOff() {
        if (isActive()) {
            return stop();
        } else {
            return start();
        }
    }

    @Override
    public boolean isActive() {
        return getReminder().isActive();
    }

    @Override
    public boolean canSkip() {
        return false;
    }

    @Override
    public boolean isRepeatable() {
        return false;
    }

    @Override
    public void setDelay(int delay) {
        if (delay == 0) {
            next();
            return;
        }
        getReminder().setDelay(delay);
        super.save();
        super.setDelay(delay);
    }

    @Override
    public long calculateTime(boolean isNew) {
        return TimeCount.getInstance(getContext()).generateDateTime(getReminder().getEventTime(), getReminder().getRepeatInterval());
    }
}

File: app/src/main/java/com/elementary/tasks/core/controller/ShoppingEvent.kt
package com.elementary.tasks.core.controller

import android.text.TextUtils
import com.github.naz013.appwidgets.AppWidgetUpdater
import com.elementary.tasks.core.services.JobScheduler
import com.elementary.tasks.core.utils.GoogleCalendarUtils
import com.elementary.tasks.core.utils.Notifier
import com.github.naz013.common.datetime.DateTimeManager
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.domain.Reminder
import com.github.naz013.common.TextProvider
import com.github.naz013.repository.GoogleTaskRepository
import com.github.naz013.repository.ReminderRepository
import org.threeten.bp.LocalDateTime

class ShoppingEvent(
  reminder: Reminder,
  reminderRepository: ReminderRepository,
  prefs: Prefs,
  googleCalendarUtils: GoogleCalendarUtils,
  notifier: Notifier,
  jobScheduler: JobScheduler,
  appWidgetUpdater: AppWidgetUpdater,
  textProvider: TextProvider,
  private val dateTimeManager: DateTimeManager,
  googleTaskRepository: GoogleTaskRepository
) : RepeatableEventManager(
  reminder,
  reminderRepository,
  prefs,
  googleCalendarUtils,
  notifier,
  jobScheduler,
  appWidgetUpdater,
  textProvider,
  dateTimeManager,
  googleTaskRepository
) {

  override val isActive: Boolean
    get() = reminder.isActive

  override fun justStart() {
    if (
      reminder.hasReminder && !TextUtils.isEmpty(reminder.eventTime) &&
      dateTimeManager.isCurrent(reminder.eventTime)
    ) {
      reminder.isActive = true
      reminder.isRemoved = false
      super.save()
      super.enableReminder()
    } else {
      reminder.isActive = true
      reminder.isRemoved = false
      super.save()
    }
  }

  override fun enable(): Boolean {
    return if (reminder.hasReminder) {
      if (!TextUtils.isEmpty(reminder.eventTime) && dateTimeManager.isCurrent(reminder.eventTime)) {
        reminder.isActive = true
        reminder.isRemoved = false
        super.save()
        super.enableReminder()
        true
      } else {
        false
      }
    } else {
      reminder.isActive = true
      reminder.isRemoved = false
      super.save()
      true
    }
  }

  override fun skip(): Boolean {
    if (canSkip()) {
      val time = dateTimeManager.generateDateTime(
        reminder.eventTime,
        reminder.repeatInterval,
        dateTimeManager.fromGmtToLocal(reminder.eventTime) ?: LocalDateTime.now()
      )
      reminder.eventTime = dateTimeManager.getGmtFromDateTime(time)
      enable()
      return true
    }
    return false
  }

  override fun next(): Boolean {
    return disable()
  }

  override fun onOff(): Boolean {
    return if (isActive) {
      disable()
    } else {
      enable()
    }
  }

  override fun canSkip(): Boolean {
    return reminder.isRepeating() && (!reminder.isLimited() || !reminder.isLimitExceed())
  }

  override fun setDelay(delay: Int) {
    if (delay == 0) {
      next()
      return
    }
    reminder.delay = delay
    super.save()
    super.setDelay(delay)
  }

  override fun calculateTime(isNew: Boolean): LocalDateTime {
    return dateTimeManager.generateDateTime(reminder.eventTime, reminder.repeatInterval)
  }
}
